# **PROYECTO: SISTEMA DE GESTIÓN DE DATOS CON PYTHON**

Una vez visto el ***Módulo 1: Fundamentos de Programación***, se presenta la siguiente actividad para que el alumnado pueda poner a prueba lo aprendido durante este módulo.

Dentro de la siguiente dinámica, realizar un **menú CRUD** donde el tema a emplear es libre. Cada función deberá ejecutar una **operación del CRUD**, agregando una opción de **visualización**:

*   **Ver todos los registros**
*   **Crear nuevo registro**
*   **Consultar registro**
*   **Actualizar registro**
*   **Eliminar registro**
*   **Salir menú**

In [ ]:
experimentos = [
    {
        "Nombre": "Doble Rendija",
        "Investigador": "Thomas Young",
        "Area": "Física",
        "Año": 1801,
        "Resultado": "Demostró la naturaleza ondulatoria de la luz"
    },
    {
        "Nombre": "Penicilina",
        "Investigador": "Alexander Fleming",
        "Area": "Biología",
        "Año": 1928,
        "Resultado": "Descubrimiento del primer antibiótico"
    },
    {
        "Nombre": "Relatividad Especial",
        "Investigador": "Albert Einstein",
        "Area": "Física",
        "Año": 1905,
        "Resultado": "Dilatación del tiempo y espacio"

    },
    {
        "Nombre": "Teoría endosimbiótica",
        "Investigador": "Lynn Margulis",
        "Area": "Biología",
        "Año": 1967,
        "Resultado": "Evolución de células eucariotas a partir de procariotas"

    }
]

In [ ]:
def ver_experimentos(experimentos):

    if not experimentos:
        print("No hay experimentos registrados.")
    else:
        for i, exp in enumerate(experimentos):
            print(f"\n========= Experimento {i+1} ===========")
            print(f"Nombre: {exp['Nombre']}")
            print(f"Investigador: {exp['Investigador']}")
            print(f"Área: {exp['Area']}")
            print(f"Año: {exp['Año']}")
            print(f"Resultado: {exp['Resultado']}")

In [ ]:
ver_experimentos(experimentos)


========= Experimento 1 ===========
Nombre: Doble Rendija
Investigador: Thomas Young
Área: Física
Año: 1801
Resultado: Demostró la naturaleza ondulatoria de la luz

========= Experimento 2 ===========
Nombre: Penicilina
Investigador: Alexander Fleming
Área: Biología
Año: 1928
Resultado: Descubrimiento del primer antibiótico

========= Experimento 3 ===========
Nombre: Relatividad Especial
Investigador: Albert Einstein
Área: Física
Año: 1905
Resultado: Dilatación del tiempo y espacio

========= Experimento 4 ===========
Nombre: Teoría endosimbiótica
Investigador: Lynn Margulis
Área: Biología
Año: 1967
Resultado: Evolución de células eucariotas a partir de procariotas


In [ ]:
def crear_experimento(experimentos):
    nombre = input("Nombre del experimento: ")
    investigador = input("Investigador: ")
    area = input("Área científica: ")
    año = input("Año: ")
    resultado = input("Resultado principal: ")

    nuevo = {
        "Nombre": nombre,
        "Investigador": investigador,
        "Area": area,
        "Año": año,
        "Resultado": resultado
    }

    experimentos.append(nuevo)
    print("\nExperimento registrado con éxito.")

In [ ]:
crear_experimento(experimentos)

Nombre del experimento: Experimento de Oersted
Investigador: Hans Christian Oersted
Área científica: 1820
Año: Fisica
Resultado principal: Relación entre la electricidad y el magnetismo

Experimento registrado con éxito.


Nombre del experimento: Experimento de Oersted
Investigador: Hans Christian Oersted
Área científica: Física
Año: 1820
Resultado principal: Relación entre la electricidad y el magnetismo

Experimento registrado con éxito

In [ ]:
def buscar_experimento(experimentos, accion):

    nombres = []
    for exp in experimentos:
        nombres.append(exp["Nombre"])

    print("----Estos son los experimentos:\n")
    print("\n".join(nombres))
    print()

    a = input(f"¿Qué experimento quieres {accion}? \n")
    print()

    for exp in experimentos:
        if exp["Nombre"].lower() == a.lower():
            return exp

    print("Experimento no encontrado.")
    return None

In [ ]:
buscar_experimento(experimentos, "ver")

----Estos son los experimentos:

Doble Rendija
Penicilina
Relatividad Especial
Teoría endosimbiótica

¿Qué experimento quieres ver? 
penicilina



{'Nombre': 'Penicilina',
 'Investigador': 'Alexander Fleming',
 'Area': 'Biología',
 'Año': 1928,
 'Resultado': 'Descubrimiento del primer antibiótico'}

In [ ]:
def actualizar_experimento(experimentos):

    while True:

        exp = buscar_experimento(experimentos, "actualizar")

        if exp is None:
            return

        while True:
            print("===== ACTUALIZAR REGISTRO =====")
            print("1. Actualizar nombre")
            print("2. Actualizar investigador")
            print("3. Actualizar área")
            print("4. Actualizar año")
            print("5. Actualizar resultado")
            print("6. Elegir otro experimento")
            print("7. Salir")
            print()
            opcion = input("Selecciona una opción: ")
            print()

            if opcion == "1":
                exp["Nombre"] = input("Nuevo nombre: ")
            elif opcion == "2":
                exp["Investigador"] = input("Nuevo investigador: ")
            elif opcion == "3":
                exp["Area"] = input("Nueva área: ")
            elif opcion == "4":
                exp["Año"] = input("Nuevo año: ")
            elif opcion == "5":
                exp["Resultado"] = input("Nuevo resultado: ")
            elif opcion == "6":
                break
            elif opcion == "7":
                return
            else:
                print("Opción no válida.")

            print("\nRegistro actualizado:", exp)
            print()

In [ ]:
actualizar_experimento(experimentos)

----Estos son los experimentos:

Doble Rendija
Penicilina
Relatividad Especial
Teoría endosimbiótica
Experimento de Millikan

¿Qué experimento quieres actualizar? 
experimento de millikan

===== ACTUALIZAR REGISTRO =====
1. Actualizar nombre
2. Actualizar investigador
3. Actualizar área
4. Actualizar año
5. Actualizar resultado
6. Elegir otro experimento
7. Salir

Selecciona una opción: 4

Nuevo año: 1812

Registro actualizado: {'Nombre': 'Experimento de Millikan', 'Investigador': 'Robert Millikan', 'Area': 'Fisica', 'Año': '1812', 'Resultado': 'Carga del electron'}

===== ACTUALIZAR REGISTRO =====
1. Actualizar nombre
2. Actualizar investigador
3. Actualizar área
4. Actualizar año
5. Actualizar resultado
6. Elegir otro experimento
7. Salir

Selecciona una opción: 4

Nuevo año: 1912

Registro actualizado: {'Nombre': 'Experimento de Millikan', 'Investigador': 'Robert Millikan', 'Area': 'Fisica', 'Año': '1912', 'Resultado': 'Carga del electron'}

===== ACTUALIZAR REGISTRO =====
1. Actual

In [ ]:
def eliminar_experimento(experimentos):

    nombres = []
    for exp in experimentos:
        nombres.append(exp["Nombre"])

    print("----Estos son los experimentos:\n")
    print("\n".join(nombres))
    print()

    nombre = input("Nombre del experimento a eliminar: ")

    for exp in experimentos:
        if exp["Nombre"].lower() == nombre.lower():
            experimentos.remove(exp)
            print("Experimento eliminado correctamente.")
            return

    print("Experimento no encontrado.")

In [ ]:
eliminar_experimento(experimentos)

----Estos son los experimentos:

Doble Rendija
Penicilina
Relatividad Especial
Teoría endosimbiótica
Experimento de Millikan

Nombre del experimento a eliminar: experimento de millikan
Experimento eliminado correctamente.


In [ ]:
def menu_experimentos(experimentos):
    while True:
        print("\n========== MENÚ DE GESTIÓN DE EXPERIMENTOS ==========")
        print()
        print("1. Ver todos los registros")
        print("2. Crear nuevo registro")
        print("3. Consultar registro")
        print("4. Actualizar registro")
        print("5. Eliminar registro")
        print("6. Salir")
        print()
        opcion = input("Selecciona una opción: ")
        print()
        if opcion == "1":
            ver_experimentos(experimentos)
        elif opcion == "2":
            crear_experimento(experimentos)
        elif opcion == "3":
            exp= buscar_experimento(experimentos, accion="consultar")
            print(exp)
        elif opcion == "4":
            actualizar_experimento(experimentos)
        elif opcion == "5":
            eliminar_experimento(experimentos)
        elif opcion == "6":
            print("Saliendo, gracias por consultar los experimentos.")
            break
        else:
            print("Opción no encontrada.")

In [ ]:
menu_experimentos(experimentos)


========== MENÚ DE GESTIÓN DE EXPERIMENTOS ==========

1. Ver todos los registros
2. Crear nuevo registro
3. Consultar registro
4. Actualizar registro
5. Eliminar registro
6. Salir

Selecciona una opción: 1


========= Experimento 1 ===========
Nombre: Doble Rendija
Investigador: Thomas Young
Área: Física
Año: 1801
Resultado: Demostró la naturaleza ondulatoria de la luz

========= Experimento 2 ===========
Nombre: Penicilina
Investigador: Alexander Fleming
Área: Biología
Año: 1928
Resultado: Descubrimiento del primer antibiótico

========= Experimento 3 ===========
Nombre: Relatividad Especial
Investigador: Albert Einstein
Área: Física
Año: 1905
Resultado: Dilatación del tiempo y espacio

========= Experimento 4 ===========
Nombre: Teoría endosimbiótica
Investigador: Lynn Margulis
Área: Biología
Año: 1967
Resultado: Evolución de células eucariotas a partir de procariotas

========== MENÚ DE GESTIÓN DE EXPERIMENTOS ==========

1. Ver todos los registros
2. Crear nuevo registro
3. Consu

Nombre del experimento: Experimento de Oersted Investigador: Hans Christian Oersted Área científica: Física Año: 1820 Resultado principal: Relación entre la electricidad y el magnetismo